### PR TITLE
core: properly reset the statedb on sethead

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This pr fixes an issue where local transactions that were included in the chain before a `SetHead` could not be replayed since the txpool still had the old statedb after the `setHead` and would reject them with `nonce to low`.
For a repro see the linked issue

resolves: https://github.com/ethereum/go-ethereum/issues/19346

### CODE REVIEW

**Review by holiman — COMMENTED** [review]
Semantically, this totally LGTM, and it's correct. 

However, it would be totally possible to rewrite this a bit, so that instead of updating the things inside of the nested clause, you reorder the clauses a bit so that this case can just fall through. That way we don't need to increase the code size, and you would also get that pool fork update fixed.

File: core/tx_pool.go [hunk 1]
```diff
@@ -1137,44 +1137,46 @@ func (pool *TxPool) reset(oldHead, newHead *types.Header) {
...
 						"old", oldHead.Hash(), "oldnum", oldNum, "new", newHead.Hash(), "newnum", newNum)
-				}
-				return
-			}
-			for rem.NumberU64() > add.NumberU64() {
-				discarded = append(discarded, rem.Transactions()...)
-				if rem = pool.chain.GetBlock(rem.ParentHash(), rem.NumberU64()-1); rem == nil {
-					log.Error("Unrooted old chain seen by tx pool", "block", oldHead.Number, "hash", oldHead.Hash())
 					return
+				} else {
```
Comment (holiman) [inline core/tx_pool.go:1145]: No need to `else` after a return

**Review by holiman — COMMENTED** [review]
LGTM, minor nitpick

**Review by holiman — APPROVED** [review]
LGTM

**Comment by unparalleled-js** [comment]
I think part of this issue is unresolved:

https://github.com/ethereum/go-ethereum/issues/26154

**Pull request merged**